Improve NumPy page (#57)
RSE-Sheffield/pando-python: episodes/optimisation-numpy.md

1. The improved code starts by extracting the `shapely.Point`s and corresponding point names as two separate NumPy arrays from a larger data frame.
2. It then passes that array of points to `current_polygon.contains()`, which uses vectorisation to speed up the calculation. It returns a NumPy array of booleans (`True` or `False`), describing for each `Point` in the input array whether it is contained in `current_polygon`.
3. This boolean array is then [passed as an index](https://numpy.org/doc/stable/user/basics.indexing.html#boolean-array-indexing) to the `point_names_list` array. This returns a new array with the names of all points that are contained in the polygon (i.e. where the boolean array had the value `True`).
4. Finally, the contained points are stored as a Python list. (In this particular case, later parts of the data analysis code expected a list instead of a NumPy array. Since those parts of the code were "fast enough"—remember Donald Knuth’s quote in the earlier episode?—the researcher decided not to spend more time to rewrite them.)

::::::::::::::::::::::::::::::::::::: instructor

The following code snippet demonstrates how this works for a simplified example.

> If you want to run this as a live demo, you need to `pip install shapely` first.

Reviewer: I meant the up-front setup notes, rather than an instructor getting caught out half way through delivery.

https://rse.shef.ac.uk/pando-python/instructor/index.html#software-setup

```Python
>>> ls = list(range(6))
>>> ls + 10
Traceback (most recent call last):
  File "<python-input-8>", line 1, in <module>
    ls + 10
    ~~~^~~~
TypeError: can only concatenate list (not "int") to list
>>> ls * 2
[0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5]
>>> ls ** 2
Traceback (most recent call last):
  File "<python-input-10>", line 1, in <module>
    ls ** 2
    ~~~^^~~
TypeError: unsupported operand type(s) for ** or pow(): 'list' and 'int'
>>> np.exp(ls)  # works but is slower, because NumPy converts the list into an array first
array([  1.        ,   2.71828183,   7.3890561 ,  20.08553692,
        54.59815003, 148.4131591 ])
```
:::::::::::::::::::::::::::::::::::::::::::::

However, broadcasting is not just a nicer way to write mathematical expressions—it can also give a significant performance boost:
Most modern processors are able to apply one instruction across multiple variables simultaneously, instead of sequentially. (In computer science, this is also referred to as "vectorisation".) The manner by which NumPy stores data in arrays enables it to vectorise mathematical operations that are broadcast across arrays.

> <!-- Analogy: If you're baking cookies, the oven (CPU register) is big enough to operate on multiple cookies (numbers) simultaneously. So whether you bake 1 cookie or 10, it'll take exactly the same amount of time. -->

Reviewer: I like it, but it's a shame it doesn't accommodate for the need for the cookies to be arranged in a specific way.

From ChatGPT (though this is too verbose, and the arrangement feels forced, I suppose I should accept an analogy will never be perfect)

> Imagine you have a pile of boxes to move, and you can only carry them one at a time—this is like processing data sequentially. It’s slow and inefficient. But if you have a dolly that lets you move multiple boxes at once, you can transport them much faster—that’s vectorization, where multiple data elements are processed simultaneously.
>
> However, for this speedup to work, the boxes need to be neatly stacked and fit properly on the dolly. If they're scattered randomly or shaped irregularly, you'll struggle to load them efficiently. Similarly, for vectorization to be effective, data needs to be structured in a way that aligns well with the underlying hardware, such as being stored in contiguous memory or in a format suited for SIMD (Single Instruction, Multiple Data) execution.

After a bit of fighting, got Claude to give me one that's concise and cares about arrangement. But the outcome of lack of arrange is rather forced

> Vectorization is like loading a dishwasher versus hand-washing. In a dishwasher, you can clean many dishes simultaneously—much faster than one-by-one hand-washing—but only if the dishes are properly arranged in rows and racks (data alignment). If dishes are stacked randomly or don't fit the predetermined arrangement pattern, the dishwasher can't clean them effectively, and you lose the speed advantage of parallel processing.

🤷‍♂️

Reviewer: I think I'm too much of an analogy perfectionist.

If you want to include it as-is, or make it into an instructor note go for it. With it being commented, it won't be much use unless people read the source.

```sh
> python -m timeit -s "import numpy; ar = numpy.arange(1)" "ar + 10"
1000000 loops, best of 5: 359 nsec per loop
> python -m timeit -s "import numpy; ar = numpy.arange(10)" "ar + 10"
1000000 loops, best of 5: 362 nsec per loop
> python -m timeit -s "import numpy; ar = numpy.arange(100)" "ar + 10"
1000000 loops, best of 5: 364 nsec per loop
```
If we were to use a regular `for` loop, the time to perform this operation would increase with the length of the array.
However, using NumPy broadcasting we can apply the addition to 1, 10 or 100 elements, all in the same amount of time!

> Earlier in this episode it was demonstrated that using core Python methods over a list will outperform a loop, performing the same calculation faster. The below example takes this a step further by demonstrating the calculation of a dot product.

Reviewer: ```suggestion
Earlier it was demonstrated that using core Python methods over a list will outperform a loop, performing the same calculation faster. The below example takes this a step further by demonstrating the calculation of a dot product.
```

In Carpentries terminology, "episode" is a page, since the splitting this is now 2 episodes earlier.

Could potentially cross link back to that episode? (I'm doing this all in GitHub, so haven't got it setup easily to test the cross link works correctly, hence no suggestions)

Author: Good catch!

:::::::::::::::::::::::::::::::::::::

> ### `vectorize()`

Reviewer: Are you sure it's worth removing this? Should it instead be reduced to a single sentence and moved to a callout.

The name overlap with a term we've introduced feels like a pitfall that people could fall into.

Author: In my experience, `np.vectorize()` is used very rarely (in fact, I don’t think I’ve encountered it “in the wild” before looking at this course), so most learners won’t come across it at all. And even if they do, the NumPy docs clearly state that this won’t give a performance boost. (And even if they manage to miss that, the worst case is that their function is about as slow as before, but slightly more convenient to use.)

I would rather not add more mental load for learners, just to try and avoid a relatively harmless issue that most of them wouldn’t ever encounter.
Maybe an instructor note instead?

Reviewer: Nah, I'll go with removing it entirely. If it does get mentioned by an attendee I'll consider adding it back.

## Other Libraries That Use NumPy

Across the scientific Python software ecosystem, [many domain-specific packages](https://numpy.org/#:~:text=ECOSYSTEM) are built on top of NumPy arrays.
Similar to the demos above, we can often gain significant performance boosts by using these libraries well.

> ::::::::::::::::::::::::::::::::::::: challenge

Reviewer: Is this more of an exercise than a challenge? (e.g. it will take time, rather than being something they can answer straight away)

- Give it a title regardless even if just "Challenge"
- If it's an exercise, assign it a duration in the title

(e.g. https://github.com/RSE-Sheffield/pando-python/blob/5725fea3ccc65b54dcc26356bc191548e31f7bb4/episodes/profiling-introduction.md?plain=1#L213)

Author: In our course, I showed the NumPy home page with the list of packages and asked learners if they already used any of these—or if any looked relevant to their work.
So … a small interactive element, but not really an exercise.

Author: Added a title in the latest commit.

Take a look at the [list of libraries on the NumPy website](https://numpy.org/#:~:text=ECOSYSTEM). Are you using any of them already?

If you've brought a project you want to work on: Are there areas of the project where you might benefit from adapting one of these libraries instead of writing your own code from scratch?

:::::::::::::::::::::::: hint

These libraries could be specific to your area of research; but they could also include packages from other fields that provide tools you need (e.g. statistics or machine learning)!

:::::::::::::::::::::::::::::::::

:::::::::::::::::::::::::::::::::::::::::::::::

> Which libraries you may use will depend on your research domain; here, we'll show two examples from our own experience.

Reviewer: ```suggestion
Which libraries you may use will depend on your research domain; here, we'll show an example from bioinformatics.
```

### Example: Image Analysis with Shapely

> A bioinformatics researcher had a large data set of images of cells. She had already reconstructed the locations of cell walls and various points of interest and needed to identify which points were located in each cell.

Reviewer: Could change this to "they" throughout or rewrite it to be more "A large data set of images of cells has been reconstructed...", but I don't think it matters strongly.

To do this, she used the [Shapely](https://github.com/shapely/shapely) geometry library.

```Python
points_per_polygon = {}
for polygon_idx in range(n_polygons):
    current_polygon = polygons.iloc[polygon_idx,:]["geometry"]

    # manually loop over all points, check if polygon contains that point
    out_points = []
    for i in range(n_points):
        current_point = points.iloc[i, :]

>         if current_polygon.contains(current_point["geometry"]):

Reviewer: ```suggestion
        # apply 'contains' to an individual point, rather than using vectorisation
        if current_polygon.contains(current_point["geometry"]):
```

To hopefully draw the contrast with the below example

Author: I can’t quite put my finger on it, but for some reason that comment doesn’t feel good to me. I think it’s because it seems to imply that we could change that particular line to vectorize the code … when really the fundamental issue is a few lines above (the for loop over points)?

I’d maybe edit the comment in l. 283 instead; although this sort of contrast (between an inefficient `for` loop and a more efficient way to achieve the same goal) is something we’ve highlighted so often already throughout the course, that I really think “manually loop over points” in the existing comment is clear.

            out_points.append(current_point.name)

    points_per_polygon[polygon_idx] = out_points
```

For about 500k points and 1000 polygons, the initial version of the code took about 20 hours to run.

Luckily, Shapely is built on top of NumPy, so she was able to apply functions to an array of points instead and wrote an improved version, which took just 20 minutes:

```Python
# Extract points and corresponding names as two separate NumPy arrays from a larger data frame
# This will make it easier to apply vectorised functions below
points_array = np.array(points.loc[:,"geometry"])
point_names_array = np.array(points.loc[:,"name"])

points_per_polygon = {}
for polygon_idx in range(n_polygons):
    current_polygon = polygons.iloc[polygon_idx,:]["geometry"]

>     # vectorised: apply `contains` to an array of points at once

Reviewer: ```suggestion
    # vectorised: apply `contains` to an array of points, rather than an individual point
```

Author: This is nice and a bit clearer, thanks!

    points_in_polygon_idx = current_polygon.contains(points_array)
    points_in_polygon = point_names_array[points_in_polygon_idx]
    
    points_per_polygon[polygon_idx] = points_in_polygon.tolist()
```
::::::::::::::::::::::::::::::::::::: instructor

To vectorise this efficiently, the logic of the code had to be changed slightly.

> The improved code starts by extracting the `shapely.Point`s and corresponding point names as two separate NumPy arrays from a larger data frame.

Reviewer: Can this be turned into a numbered list, with the numbers annotated via comments in the code?

Reviewer: Maybe also move it above the second code block, as it's probably easier to parse the code in hindsight of reading this. Although then the simplified example doesn't fit. 🤷‍♂️

Reviewer: I definitely think what's visible to attendees needs more explanation. So maybe move the list up (and out of the instructor box) and leave the simple demonstration in the instructor box.

It may also be worth adding `shapely` to the setup notes in an instructor box referring to this episode.

Author: Good points; I’ll edit this.
